test/api/v3/unit/models/challenge.test.js

```diff
@@ -0,0 +1,154 @@
+        let updatedNewMemberTasks = await Tasks.Task.find({_id: { $in: updatedNewMember.tasksOrder[`${taskType}s`]}});
+        let syncedTask = find(updatedNewMemberTasks, function findNewTask (updatedNewMemberTask) {
+          return updatedNewMemberTask.type === taskValue.type && updatedNewMemberTask.text === taskValue.text;
+        });
+
+        expect(updatedNewMember.challenges).to.contain(challenge._id);
+        expect(updatedNewMember.tags[3]._id).to.equal(challenge._id);
+        expect(updatedNewMember.tags[3].name).to.equal(challenge.shortName);
+        expect(syncedTask).to.exist;
+      });
+
+      it('updates tasks to challenge and challenge members', async () => {
+        let updatedTaskName = 'Updated Test Habit';
+        await challenge.addTasks([task]);
+
+        _.assign(task, _.merge(task.toObject(), Tasks.Task.sanitizeUpdate({ text: updatedTaskName })));
+        await challenge.updateTask(task);
+
+        let updatedLeader = await User.findOne({_id: leader._id});
+        let updatedUserTask = await Tasks.Task.findById(updatedLeader.tasksOrder[`${taskType}s`][0]);
+
+        expect(updatedUserTask.text).to.equal(updatedTaskName);
+      });
+
+      it('removes a tasks to challenge and challenge members', async () => {
+        await challenge.addTasks([task]);
+        await challenge.removeTask(task);
+
+        let updatedLeader = await User.findOne({_id: leader._id});
+        let updatedUserTask = await Tasks.Task.findOne({_id: updatedLeader.tasksOrder[`${taskType}s`][0]}).exec();
+
+        expect(updatedUserTask.challenge.broken).to.equal('TASK_DELETED');
+      });
+
+      it('unlinks and deletes challenge tasks for a user when remove-all is specified', async () => {
+        await challenge.addTasks([task]);
+        await challenge.unlinkTasks(leader, 'remove-all');
+
+        let updatedLeader = await User.findOne({_id: leader._id});
+        let updatedLeadersTasks = await Tasks.Task.find({_id: { $in: updatedLeader.tasksOrder[`${taskType}s`]}});
+        let syncedTask = find(updatedLeadersTasks, function findNewTask (updatedLeadersTask) {
+          return updatedLeadersTask.type === taskValue.type && updatedLeadersTask.text === taskValue.text;
+        });
+
+        expect(syncedTask).to.not.exist;
+      });
+
+      it('unlinks and keeps challenge tasks for a user when keep-all is specified', async () => {
+        await challenge.addTasks([task]);
+        await challenge.unlinkTasks(leader, 'keep-all');
+
+        let updatedLeader = await User.findOne({_id: leader._id});
+        let updatedLeadersTasks = await Tasks.Task.find({_id: { $in: updatedLeader.tasksOrder[`${taskType}s`]}});
+        let syncedTask = find(updatedLeadersTasks, function findNewTask (updatedLeadersTask) {
+          return updatedLeadersTask.type === taskValue.type && updatedLeadersTask.text === taskValue.text;
+        });
+
+        expect(syncedTask).to.exist;
+        // expect(syncedTask.challenge).to.be.empty;
```

Comment:
Hmmm. This is what is shows:

```
AssertionError: expected {} to deeply equal {}
      + expected - actual

      -{
      -  "$__": {
      -    "_id": [undefined]
      -    "activePaths": {
      -      "paths": {
      -        "__v": "init"
      -        "_id": "init"
      -        "attribute": "init"
      -        "createdAt": "init"
      -        "notes": "init"
      -        "priority": "init"
      -        "tags": "init"
      -        "text": "init"
      -        "type": "init"
      -        "updatedAt": "init"
      -        "userId": "init"
      -        "value": "init"
      -      }
      -      "stateNames": [
      -        "require"
      -        "modify"
      -        "init"
      -        "default"
      -        "ignore"
      -      ]
      -      "states": {
      -        "default": {}
      -        "ignore": {}
      -        "init": {
      -          "__v": true
      -          "_id": true
      -          "attribute": true
      -          "createdAt": true
      -          "notes": true
      -          "priority": true
      -          "tags": true
      -          "text": true
      -          "type": true
      -          "updatedAt": true
      -          "userId": true
      -          "value": true
      -        }
      -        "modify": {}
      -        "require": {}
      -      }
      -    }
      -    "adhocPaths": [undefined]
      -    "emitter": {
      -      "_events": {}
      -      "_eventsCount": 0
      -      "_maxListeners": 0
      -      "domain": [null]
      -    }
      -    "fullPath": [undefined]
      -    "getters": {
      -      "challenge": [Circular]
      -    }
      -    "inserting": [undefined]
      -    "ownerDocument": [undefined]
      -    "populate": [undefined]
      -    "populated": [undefined]
      -    "removing": [undefined]
      -    "saveError": [undefined]
      -    "scope": {
      -      "$__": [Circular]
      -      "$__original_save": {
      -        "numAsyncPres": 1
      -      }
      -      "_doc": {
      -        "__v": 0
      -        "_id": "dd667e40-e464-4fd3-9919-f77f0d597c28"
      -        "attribute": "str"
      -        "challenge": {}
      -        "createdAt": [Date: 2016-03-04T00:23:17.163Z]
      -        "notes": ""
      -        "priority": 1
      -        "tags": []
      -        "text": "test reward"
      -        "type": "reward"
      -        "updatedAt": [Date: 2016-03-04T00:23:17.163Z]
      -        "userId": "8efdc19e-ae90-4e4c-aeb7-278a2afcf46a"
      -        "value": 0
      -      }
      -      "_posts": {
      -        "$__original_save": []
      -      }
      -      "_pres": {
      -        "$__original_save": [
      -          {
      -            "isAsync": false
      -          }
      -          {
      -            "isAsync": true
      -          }
      -          {
      -            "isAsync": false
      -          }
      -        ]
      -      }
      -      "errors": [undefined]
      -      "isNew": false
      -      "save": [Function]
      -    }
      -    "selected": [undefined]
      -    "shardval": [undefined]
      -    "strictMode": true
      -    "validationError": [undefined]
      -    "version": [undefined]
      -    "wasPopulated": false
      -  }
      -  "$__isNested": true
      -  "$__original_save": {
      -    "numAsyncPres": 1
      -  }
      -  "_doc": {
      -    "__v": 0
      -    "_id": "dd667e40-e464-4fd3-9919-f77f0d597c28"
      -    "attribute": "str"
      -    "challenge": {}
      -    "createdAt": [Date: 2016-03-04T00:23:17.163Z]
      -    "notes": ""
      -    "priority": 1
      -    "tags": []
      -    "text": "test reward"
      -    "type": "reward"
      -    "updatedAt": [Date: 2016-03-04T00:23:17.163Z]
      -    "userId": "8efdc19e-ae90-4e4c-aeb7-278a2afcf46a"
      -    "value": 0
      -  }
      -  "_posts": {
      -    "$__original_save": []
      -  }
      -  "_pres": {
      -    "$__original_save": [
      -      {
      -        "isAsync": false
      -      }
      -      {
      -        "isAsync": true
      -      }
      -      {
      -        "isAsync": false
      -      }
      -    ]
      -  }
      -  "broken": [undefined]
      -  "errors": [undefined]
      -  "id": [undefined]
      -  "isNew": false
      -  "save": [Function]
      -  "taskId": [undefined]
      -  "toJSON": [Function]
      -  "toObject": [Function]
      -  "winner": [undefined]
      -}
      +{}
```

I don't see those properties when I log to console. Are they part of Mongo?